src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,431 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import javax.crypto.KDFSpi;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.KDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * KeyDerivation implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC 5869.  This
+ * implementation provides the complete Extract-then-Expand HKDF function as well as Extract-only
+ * and Expand-only variants.
+ */
+abstract class HkdfKeyDerivation extends KDFSpi {
+
+    protected Mac hmacObj;
+    protected int hmacLen;
+    protected String hmacAlgName;
+    protected List<SecretKey> ikms;
+    protected List<SecretKey> salts;
+    protected SecretKey initialKeyMaterial;
+    protected SecretKey salt;
+    protected SecretKey pseudoRandomKey;
+    protected byte[] info;
+    protected int length;
+
+    protected enum HKDF_TYPE {
+        EXTRACT, EXPAND, EXTRACTEXPAND
+    }
+
+    protected HKDF_TYPE hkdfType;
+
+    /**
+     * The sole constructor.
+     *
+     * @param algParameterSpec
+     *     the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this {@code KDFSpi}
+     */
+    protected HkdfKeyDerivation(AlgorithmParameterSpec algParameterSpec)
```

Comment:
I suggest throwing an exception if params is not null.

Reply:
The params is provided through `getInstance(alg, params)`. Just throw `InvalidAlgorithmParameterException` as specified there.